at86rf2xx: add NETOPT_TX_RETRIES_NEEDED support

bergzand opened the pull request:
Support for `NETOPT_TX_RETRIES_NEEDED` for the at86rf233. As far as I've found, only the at86rf233 supports reporting no ack retransmissions. This is based on #7448.

The code compiles, but as I don't have access to an at86rf233, I have no way to check if the code is functionally correct.

gebart commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -561,6 +566,10 @@ static void _isr(netdev_t *netdev)
                 at86rf2xx_set_state(dev, dev->idle_state);
                 DEBUG("[at86rf2xx] return to state 0x%x\n", dev->idle_state);
             }
+#ifdef MODULE_AT86RF233
```
A quick scan of the manual tells me that the RF232 provides this register as well. There is no mention of ARET_FRAME_RETRIES in the data sheets for RF212B or RF231 though.

bergzand replied:
I thought I checked them all, but thanks for checking. I must have missed the RF232.

gebart commented on drivers/at86rf2xx/include/at86rf2xx_registers.h:
```diff
@@ -321,6 +324,17 @@ extern "C" {
 /** @} */
 
 /**
+ * @brief   Bitfield definitions for the XAH_CTRL_1 register
```
`@brief` -> `@name`

gebart commented on drivers/include/at86rf2xx.h:
```diff
@@ -167,6 +167,10 @@ typedef struct {
                                              this is required to know when to
                                              return to @ref at86rf2xx_t::idle_state */
+#ifdef MODULE_AT86RF233
+    /* Only AT86RF233 supports retransmission reporting */
```
and 232

bergzand commented:
@gebart I've added the at86rf232 to the support.

For testing this functionality with the mrf24j40 I've used #6873, but I think I need to rebase that PR a bit before it can be used here without any merge conflicts. With #6873, ETX statistics can be viewed with the `ifconfig` command. Correct functionality is assumed when the ETX can be varied between 1.00 and NETSTATS_NB_ETX_NOACK_PENALTY, based on radio output power or whatever is used to manipulate the link quality 😃.

gebart commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -561,6 +568,10 @@ static void _isr(netdev_t *netdev)
                 at86rf2xx_set_state(dev, dev->idle_state);
                 DEBUG("[at86rf2xx] return to state 0x%x\n", dev->idle_state);
             }
+#if defined(MODULE_AT86RF232) || defined(MODULE_AT86RF233)
```
Optionally you could change these checks to ifdef AT86RF2XX_XAH_CTRL_2__ARET_FRAME_RETRIES_OFFSET
and only have the model specific checks inside the register header. This way it will be easy to add support for a future at86rf234 or whatever

bergzand replied:
That sounds like a good idea

bergzand commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -561,6 +568,10 @@ static void _isr(netdev_t *netdev)
             }
+#if defined(MODULE_AT86RF232) || defined(MODULE_AT86RF233)
+            dev->tx_retries = at86rf2xx_reg_read(dev, AT86RF2XX_REG__XAH_CTRL_2) >>
+                              AT86RF2XX_XAH_CTRL_2__ARET_FRAME_RETRIES_OFFSET;
```
Yes, I think that's a bit better too, makes it easier if anything changes with future at86rf versions too

smlng commented on drivers/include/at86rf2xx.h:
```diff
@@ -167,6 +167,11 @@ typedef struct {
     uint8_t pending_tx;                 /**< keep track of pending TX calls
                                              this is required to know when to
                                              return to @ref at86rf2xx_t::idle_state */
+#if defined(MODULE_AT86RF232) || defined(MODULE_AT86RF233)
```
IMO the condition here should match the above, i.e., `AT86RF2XX_XAH_CTRL_2__ARET_FRAME_RETRIES_OFFSET`. Just to make sure its consistent and avoid failures in the future.

bergzand replied:
Missed that one :(

bergzand commented:
@smlng @gebart With the adjusted ifdefs, I think it is less obvious which radio supports the frame retry reporting. Should there be a comment somewhere about that (for now) only the 232 and the 233 support frame retry reporting?

smlng commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -573,6 +581,11 @@ static void _isr(netdev_t *netdev)
                 at86rf2xx_set_state(dev, dev->idle_state);
                 DEBUG("[at86rf2xx] return to state 0x%x\n", dev->idle_state);
             }
+#ifdef AT86RF2XX_XAH_CTRL_2__ARET_FRAME_RETRIES_OFFSET
```
add `/* Only radios with the XAH_CTRL_2 register support frame retry reporting */` here, too

smlng commented on drivers/at86rf2xx/include/at86rf2xx_registers.h:
```diff
@@ -330,6 +333,19 @@ extern "C" {
 /** @} */
 
 /**
+ * @name    Bitfield definitions for the XAH_CTRL_1 register
```
maybe add some more detailed doc here, to clarify as you suggested yourself.

smlng commented:
feel free to add docu as suggested above, and squash directly afterwards. 

IMHO this looks (close to) ready for merging.

bergzand commented:
Documentation added and squashed.

smlng commented on drivers/include/at86rf2xx.h:
```diff
@@ -167,6 +167,11 @@ typedef struct {
                                              this is required to know when to
                                              return to @ref at86rf2xx_t::idle_state */
+/* Only radios with the XAH_CTRL_2 register support frame retry reporting */
+#ifdef AT86RF2XX_XAH_CTRL_2__ARET_FRAME_RETRIES_OFFSET
```
Murdock fails, I guess this files missed `#include "at86rf2xx_registers.h"`

bergzand replied:
Oh, of course :(

gebart replied:
I think it would be better to leave this at if MODULE_AT86RF233 etc, instead of including the internal header in the external API declarations.

bergzand commented:
Murdock shows all green now.

gebart commented on drivers/include/at86rf2xx.h:
```diff
@@ -32,6 +32,7 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+#include "at86rf2xx_registers.h"
```
I don't like this, it breaks the separation of the public API and the internal register addresses.

smlng commented:
mhm, but using `MODULE_AT86RF233` for the check, has 2 drawbacks: a) inconsistent with the other ifdef, and b) needs to be adapted for every (future) driving having this feature - making errors more likely.

gebart commented:
What I meant is that this adds a dependency on an internal header (drivers/at86rf2xx/include) from a global API header (drivers/include). This will break the separation between public API declaration and internal implementation details.

smlng commented:
@gebart I understood what you mean, and I agree. However, I also dislike the implications it has on this driver and future adaptions. From my point of view both are ugly in some way. 

A solution would be to have an intermediate define in the global header that activates that feature and can be used for all `#ifdef`s and would require only changes at one point and not several.

Something like:

```
#if defined(MODULE_AT86RF232) || defined(MODULE_AT86RF233)
#define AT86RF2XX_HAVE_RETRIES
#endif
```

in global `drivers/include/at86rf2xx.h` and use `AT86RF2XX_HAVE_RETRIES` as a condition replacing `#if defined(MODULE_AT86RF232) || defined(MODULE_AT86RF233)` or `#ifdef AT86RF2XX_XAH_CTRL_2__ARET_FRAME_RETRIES_OFFSET` throughout the code.

That way, future enhancements to newer modules would only have to add the distinct driver module in the above `#if defined ...` rule.

gebart commented:
@smlng your suggested solution looks good to me :+1:

bergzand commented:
@smlng @gebart That looks like a better solution. Give me a minute to implement and test it.

bergzand commented:
I've adjusted the code as suggested above. I've also added a few `if defined(DOXYGEN)` here and there to include the documentation in a relevant way in the doxygen files.

smlng commented on drivers/include/at86rf2xx.h:
```diff
@@ -92,6 +92,21 @@ extern "C" {
+ * contains the frame retry counter. Only the at86rf232 and the at86rf233
+ * support this register.
+ *
+ * @{
```
you don't need a doc grouping here, move the docu directly above the `#define AT86RF2XX_HAVE_RETRIES`, i.e., within the `#if ... #endif`. Thus, please remove the doc grouping `@{`  and `/** @} */`, IMO not needed here, as it is only one define.

smlng commented:
changes look good to me, minor nit see above.

bergzand commented:
@smlng: Addressed minor nit :)

smlng commented:
I re-approve, @gebart your call ...

gebart commented on drivers/include/at86rf2xx.h:
```diff
@@ -91,6 +91,18 @@ extern "C" {
+ * contains the frame retry counter. Only the at86rf232 and the at86rf233
+ * support this register.
+ */
+#define AT86RF2XX_HAVE_RETRIES
```
define this to 1 instead of just empty, makes it possible to check it in compiler code as well and not only with the preprocessor.

kaspar030 replied:
doesn't an empty define implicitly define it to "1"?

gebart replied:
@kaspar030 from the command line, yes (`-DMYDEF` is the same as `-DMYDEF=1`), but I didn't think that was the case for the preprocessor defines.

gebart replied:
Please correct me if I am wrong

bergzand replied:
I can't find any documentation to confirm @kaspar030 remark.

@gebart If you want to use the define in code as well, wouldn't it be necessary to define it as 0 by default and change it to 1 if a supported radio is enabled?

gebart replied:
@bergzand you're right, could you adjust this to have an `#else, #define AT86RF2XX_HAVE_RETRIES 0`, and change the `#ifdef` to `#if`?

gebart commented on drivers/at86rf2xx/include/at86rf2xx_registers.h:
```diff
@@ -330,6 +335,23 @@ extern "C" {
+ *
+ * @{
+ */
+#if defined(AT86RF2XX_HAVE_RETRIES) || defined(DOXYGEN)
```
defined(DOXYGEN) shouldn't really be needed since the flag is already defined if DOXYGEN is defined.

bergzand replied:
Good one, will fix in the next push

bergzand commented:
And another update.

smlng commented:
my ACK holds, @gebart happy, too?

smlng commented:
ACK and GO then
